File: DataSheets/FlowPathwayDataSheet.vb
'*********************************************************************************************
' STSimStockFlow: A SyncroSim Module for the ST-Sim Stocks and Flows Add-In.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core
Imports System.Globalization
Imports System.Reflection

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class FlowPathwayDataSheet
    Inherits DataSheet

    Protected Overrides Sub OnDataSheetChanged(e As DataSheetMonitorEventArgs)

        MyBase.OnDataSheetChanged(e)

        Dim p As String = e.GetValue("PrimaryStratumLabel", "Stratum")

        Me.Columns(FROM_STRATUM_ID_COLUMN_NAME).DisplayName = String.Format(CultureInfo.InvariantCulture, "From {0}", p)
        Me.Columns(TO_STRATUM_ID_COLUMN_NAME).DisplayName = String.Format(CultureInfo.InvariantCulture, "To {0}", p)

    End Sub

    Public Overrides Sub Validate(proposedRow As DataRow, transferMethod As DataTransferMethod)

        MyBase.Validate(proposedRow, transferMethod)

        Dim DiagramSheet As DataSheet = Me.GetDataSheet(DATASHEET_FLOW_PATHWAY_DIAGRAM_NAME)
        Dim StockTypes As Dictionary(Of Integer, Boolean) = CreateRecordLookup(DiagramSheet, STOCK_TYPE_ID_COLUMN_NAME)

        Dim FromStockTypeId As Integer = CInt(proposedRow(FROM_STOCK_TYPE_ID_COLUMN_NAME))

        If (Not StockTypes.ContainsKey(FromStockTypeId)) Then
            Throw New DataException("The 'From Stock' does not exist for this scenario.")
        End If

        Dim ToStockTypeId As Integer = CInt(proposedRow(TO_STOCK_TYPE_ID_COLUMN_NAME))

        If (Not StockTypes.ContainsKey(ToStockTypeId)) Then
            Throw New DataException("The 'To Stock' does not exist for this scenario.")
        End If

    End Sub

    Public Overrides Sub Validate(proposedData As DataTable, transferMethod As DataTransferMethod)

        MyBase.Validate(proposedData, transferMethod)

        Dim StockTypeSheet As DataSheet = Me.Project.GetDataSheet(DATASHEET_STOCK_TYPE_NAME)
        Dim DiagramSheet As DataSheet = Me.GetDataSheet(DATASHEET_FLOW_PATHWAY_DIAGRAM_NAME)
        Dim StockTypes As Dictionary(Of Integer, Boolean) = CreateRecordLookup(DiagramSheet, STOCK_TYPE_ID_COLUMN_NAME)

        For Each dr As DataRow In proposedData.Rows

            Dim FromStockTypeId As Integer = CInt(dr(FROM_STOCK_TYPE_ID_COLUMN_NAME))

            If (Not StockTypes.ContainsKey(FromStockTypeId)) Then

                Dim StockTypeName As String = CStr(DataTableUtilities.GetTableValue(StockTypeSheet.GetData(), StockTypeSheet.ValueMember, FromStockTypeId, StockTypeSheet.DisplayMember))
                Throw New DataException(String.Format(CultureInfo.InvariantCulture, "Cannot import flow pathways because the 'From Stock' does not exist in this scenario: {0}", StockTypeName))

            End If

            Dim ToStockTypeId As Integer = CInt(dr(TO_STOCK_TYPE_ID_COLUMN_NAME))

            If (Not StockTypes.ContainsKey(ToStockTypeId)) Then

                Dim StockTypeName As String = CStr(DataTableUtilities.GetTableValue(StockTypeSheet.GetData(), StockTypeSheet.ValueMember, ToStockTypeId, StockTypeSheet.DisplayMember))
                Throw New DataException(String.Format(CultureInfo.InvariantCulture, "Cannot import flow pathways because the 'To Stock' does not exist in this scenario: {0}", StockTypeName))

            End If

        Next

    End Sub

End Class

File: src/DataSheets/FlowPathwayDataSheet.cs
// stsim-stockflow: SyncroSim Add-On Package (to stsim) for integrating stocks and flows into state-and-transition simulation models in ST-Sim.
// Copyright © 2007-2021 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using SyncroSim.Core;
using System.Reflection;
using System.Globalization;
using System.Collections.Generic;

namespace SyncroSim.STSimStockFlow
{
	[ObfuscationAttribute(Exclude=true, ApplyToMembers=false)]
	internal class FlowPathwayDataSheet : DataSheet
	{
		protected override void OnDataSheetChanged(DataSheetMonitorEventArgs e)
		{
			base.OnDataSheetChanged(e);

			string psl = e.GetValue("PrimaryStratumLabel", "Stratum");
			string ssl = e.GetValue("SecondaryStratumLabel", "Secondary Stratum");
			string tsl = e.GetValue("TertiaryStratumLabel", "Tertiary Stratum");

			this.Columns[Constants.FROM_STRATUM_ID_COLUMN_NAME].DisplayName = string.Format(CultureInfo.InvariantCulture, "From {0}", psl);
			this.Columns[Constants.FROM_SECONDARY_STRATUM_ID_COLUMN_NAME].DisplayName = string.Format(CultureInfo.InvariantCulture, "From {0}", ssl);
			this.Columns[Constants.FROM_TERTIARY_STRATUM_ID_COLUMN_NAME].DisplayName = string.Format(CultureInfo.InvariantCulture, "From {0}", tsl);

			this.Columns[Constants.TO_STRATUM_ID_COLUMN_NAME].DisplayName = string.Format(CultureInfo.InvariantCulture, "To {0}", psl);

			this.Columns[Constants.TRANSFER_TO_STRATUM_ID_COLUMN_NAME].DisplayName = string.Format(CultureInfo.InvariantCulture, "Transfer to {0}", psl);
			this.Columns[Constants.TRANSFER_TO_SECONDARY_STRATUM_ID_COLUMN_NAME].DisplayName = string.Format(CultureInfo.InvariantCulture, "Transfer to {0}", ssl);
			this.Columns[Constants.TRANSFER_TO_TERTIARY_STRATUM_ID_COLUMN_NAME].DisplayName = string.Format(CultureInfo.InvariantCulture, "Transfer to {0}", tsl);
		}

		public override void Validate(DataRow proposedRow, DataTransferMethod transferMethod)
		{
			base.Validate(proposedRow, transferMethod);

			DataSheet DiagramSheet = this.GetDataSheet(Constants.DATASHEET_FLOW_PATHWAY_DIAGRAM_NAME);
			Dictionary<int, bool> StockTypes = LookupKeyUtilities.CreateRecordLookup(DiagramSheet, Constants.STOCK_TYPE_ID_COLUMN_NAME);
			int FromStockTypeId = Convert.ToInt32(proposedRow[Constants.FROM_STOCK_TYPE_ID_COLUMN_NAME], CultureInfo.InvariantCulture);

			if (!StockTypes.ContainsKey(FromStockTypeId))
			{
				throw new DataException("The 'From Stock' does not exist for this scenario.");
			}

			int ToStockTypeId = Convert.ToInt32(proposedRow[Constants.TO_STOCK_TYPE_ID_COLUMN_NAME], CultureInfo.InvariantCulture);

			if (!StockTypes.ContainsKey(ToStockTypeId))
			{
				throw new DataException("The 'To Stock' does not exist for this scenario.");
			}
		}

		public override void Validate(DataTable proposedData, DataTransferMethod transferMethod)
		{
			base.Validate(proposedData, transferMethod);

			DataSheet StockTypeSheet = this.Project.GetDataSheet(Constants.DATASHEET_STOCK_TYPE_NAME);
			DataSheet DiagramSheet = this.GetDataSheet(Constants.DATASHEET_FLOW_PATHWAY_DIAGRAM_NAME);
			Dictionary<int, bool> StockTypes = LookupKeyUtilities.CreateRecordLookup(DiagramSheet, Constants.STOCK_TYPE_ID_COLUMN_NAME);

			foreach (DataRow dr in proposedData.Rows)
			{
				int FromStockTypeId = Convert.ToInt32(dr[Constants.FROM_STOCK_TYPE_ID_COLUMN_NAME], CultureInfo.InvariantCulture);

				if (!StockTypes.ContainsKey(FromStockTypeId))
				{
					string StockTypeName = Convert.ToString(DataTableUtilities.GetTableValue(StockTypeSheet.GetData(), StockTypeSheet.ValueMember, FromStockTypeId, StockTypeSheet.DisplayMember), CultureInfo.InvariantCulture);
					throw new DataException(string.Format(CultureInfo.InvariantCulture, "Cannot import flow pathways because the 'From Stock' does not exist in this scenario: {0}", StockTypeName));
				}

				int ToStockTypeId = Convert.ToInt32(dr[Constants.TO_STOCK_TYPE_ID_COLUMN_NAME], CultureInfo.InvariantCulture);

				if (!StockTypes.ContainsKey(ToStockTypeId))
				{
					string StockTypeName = Convert.ToString(DataTableUtilities.GetTableValue(StockTypeSheet.GetData(), StockTypeSheet.ValueMember, ToStockTypeId, StockTypeSheet.DisplayMember), CultureInfo.InvariantCulture);
					throw new DataException(string.Format(CultureInfo.InvariantCulture, "Cannot import flow pathways because the 'To Stock' does not exist in this scenario: {0}", StockTypeName));
				}
			}
		}
	}
}